srcCpp/Infrastructure_common.h

```diff
@@ -36,19 +27,169 @@
 #endif
 
 /*
- * Since std::to_string is not defined until c++11
- * we will define it here.
+ * Define the units for time when using Usec or Nsec
  */
-namespace std {
-    template<typename T>
-    std::string to_string(const T &n) {
-        std::ostringstream s;
-        s << n;
-        return s.str();
+#ifndef RTI_PERFTEST_NANO_CLOCK
+  #define PERFT_TIME_UNIT "us"
+#else
+  #define PERFT_TIME_UNIT "ns"
+#endif
+
+/*
+ * Connext DDS Pro and Micro have their own implementation for:
+ *
+ * - Semaphore
+ * - Mutex
+ * - Clock
+ * - Threads
+ *
+ * Therefore we will use those, to ensure not having to use the C++11 code,
+ * which will make us compatible with less OSs. In other cases we do need a
+ * implementation for these classes.
+ */
+#ifdef PERFTEST_RTI_PRO
+  #include "Infrastructure_pro.h"
+#elif PERFTEST_RTI_MICRO
+  #include "Infrastructure_micro.h"
+#else // Other DDS Middleware
+
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+/*
+ * TODO
+ * We will keep this here for now, as depending on the DDS implementation, the
+ * generated variable for a string is different. We need to move it to a generic
+ * place and remove it from here in the future.
+ */
+
+static const char *const THROUGHPUT_TOPIC_NAME = "Throughput";
+static const char *const LATENCY_TOPIC_NAME = "Latency";
+static const char *const ANNOUNCEMENT_TOPIC_NAME = "Announcement";
+
+
+/********************************************************************/
+/* Perftest Semaphore class */
+
+#define PERFTEST_SEMAPHORE_COUNT 1
+#define PERFTEST_SEMAPHORE_TIMEOUT_INFINITE -1
+
+class PerftestSemaphore
+{
+private:
+    unsigned int count;
+    std::mutex mutex;
+    std::condition_variable condition;
+
+public:
+    inline PerftestSemaphore(unsigned int count) : count(count)
+    {
     }
-}
 
-/* Perftest Timer Class */
+    inline bool take()
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        // This is true if the managed mutex object was locked (or adopted)
+        // by the unique_lock object, and hasn't been unlocked or released
+        // since. In all other cases, it is false.
+        condition.wait(lock, [&]() -> bool { return count > 0; });
+        --count;
+        return (bool) lock;
+    }
+
+    inline bool give()
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        ++count;
+        condition.notify_one();
+        return (bool) lock;
+    }
+};
+
+PerftestSemaphore *PerftestSemaphore_new(unsigned int count);
```

Comment:
The reason why we use function instead of class methods (the right way in my opinion) is that we want avoid modifying the infrastructure_pro and infrastructure_micro and call directly the functions, with no intermediate wrappers.